File: backend/app/schemas/query.py
from __future__ import annotations

from datetime import date
from typing import Any

from pydantic import BaseModel


class QueryColumn(BaseModel):
    """查询表格列定义。

    创建日期：2026-05-04
    author: sunshengxian
    """

    key: str
    label: str
    width: int | None = None


class QueryDatasetInfo(BaseModel):
    """可查询数据集信息。

    创建日期：2026-05-04
    author: sunshengxian
    """

    name: str
    label: str
    description: str
    date_field: str | None = None
    columns: list[QueryColumn]


class DataQueryResponse(BaseModel):
    """统一数据查询响应。

    创建日期：2026-05-04
    author: sunshengxian
    """

    dataset: str
    total: int
    page: int
    page_size: int
    columns: list[QueryColumn]
    rows: list[dict[str, Any]]


class DataQueryParams(BaseModel):
    """统一数据查询参数。

    创建日期：2026-05-04
    author: sunshengxian
    """

    dataset: str
    keyword: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    page: int = 1
    page_size: int = 30

File: frontend/src/api/query.ts
import { requestJson } from './client';
import type { DataQueryParams, DataQueryResponse, QueryDatasetInfo } from '../types/domain';

/**
 * 查询可查看的数据集。
 * 创建日期：2026-05-04
 * author: sunshengxian
 */
export function fetchQueryDatasets() {
  return requestJson<QueryDatasetInfo[]>('/api/query/datasets');
}

/**
 * 查询指定数据集的同步数据。
 * 创建日期：2026-05-04
 * author: sunshengxian
 */
export function fetchQueryRows(params: DataQueryParams) {
  const search = new URLSearchParams();
  search.set('dataset', params.dataset);
  search.set('page', String(params.page));
  search.set('page_size', String(params.page_size));
  if (params.keyword) {
    search.set('keyword', params.keyword);
  }
  if (params.start_date) {
    search.set('start_date', params.start_date);
  }
  if (params.end_date) {
    search.set('end_date', params.end_date);
  }
  return requestJson<DataQueryResponse>(`/api/query/rows?${search.toString()}`);
}
